### PR TITLE
Add RolesScreen list view

### PR DESCRIPTION
**Pull request by Ngkaokis**
ref #3903 

Issue: Tried to adjust `flexGrow` of other columns but it's still always only the first column's width is adjusted

image

### CODE REVIEW

**Comment by Ngkaokis** [comment]
> What about setting maxWidth?

That's work if we don't want to adjust columns base on proportion. Let me confirm with designer

File: pkg/admin/facade/rolesgroups.go [hunk 1]
```diff
@@ -80,9 +82,13 @@ func (f *RolesGroupsFacade) ListRoles(options *rolesgroups.ListRolesOptions, pag
 		return nil, nil, err
 	}
 
+	count, err := f.RolesGroupsQueries.CountRoles()
+	if err != nil {
+		return nil, nil, err
+	}
+
```
Comment (tung2744) [inline pkg/admin/facade/rolesgroups.go:89]: ListRoles supports `SearchKeyword`, so CountRoles should also support it?

File: pkg/admin/facade/rolesgroups.go [hunk 2]
```diff
@@ -119,9 +125,13 @@ func (f *RolesGroupsFacade) ListGroups(options *rolesgroups.ListGroupsOptions, p
 		return nil, nil, err
 	}
 
+	count, err := f.RolesGroupsQueries.CountGroups()
```
Comment (tung2744) [inline pkg/admin/facade/rolesgroups.go:128]: Same as roles, it should support SearchKeyword?

File: portal/src/graphql/adminapi/RolesScreen.module.css [hunk 3]
```diff
@@ -1,12 +1,19 @@
 .content {
-  @apply mt-11;
+  height: 100%;
```
Comment (tung2744) [inline portal/src/graphql/adminapi/RolesScreen.module.css:2]: This element's parent is a flex container, therefore use flex's method to layout. Use `flex: 1 1 0px` here in desktop, and `flex: 1 0 auto` in mobile.

File: portal/src/graphql/adminapi/RolesScreen.tsx [hunk 4]
```diff
@@ -64,36 +105,92 @@ const RolesScreen: React.VFC = function RolesScreen() {
...
-        <NavBreadcrumb className="block" items={items} />
-      </div>
-      {isEmpty ? (
-        <RolesScreenEmptyState className={styles.widget} />
-      ) : (
-        // TODO: implement list view
-        <RolesScreenEmptyState className={styles.widget} />
-      )}
-    </ScreenContent>
+    <LocalSearchBoxContext.Provider value={localSearchBoxProps}>
```
Comment (tung2744) [inline portal/src/graphql/adminapi/RolesScreen.tsx:163]: What is the reason we need this provider, instead of passing the props directly?

Comment (tung2744) [reply portal/src/graphql/adminapi/RolesScreen.tsx:163]: And also the word `Local` make me think it is searching in local, but it seems not. The search is performed in server side.

**Comment by tung2744** [comment]
Screenshot 2024-03-08 at 12 02 54 PM

It seems the table is too long in narrow screens.

File: portal/src/graphql/adminapi/RolesScreen.tsx [hunk 5]
```diff
@@ -64,36 +105,92 @@ const RolesScreen: React.VFC = function RolesScreen() {
...
+      <ScreenContent className={styles.content} layout="list">
+        <div className={styles.widget}>
+          <div className={styles.titleContainer}>
+            <NavBreadcrumb className="block" items={items} />
+            {!isEmpty ? <CreateRoleButton /> : null}
+          </div>
+          <LocalSearchBox />
```
Comment (tung2744) [inline portal/src/graphql/adminapi/RolesScreen.tsx:170]: According to figma, search box shouldn't exist when empty state is showing.

File: portal/src/graphql/adminapi/RolesScreen.tsx [hunk 6]
```diff
@@ -64,36 +105,92 @@ const RolesScreen: React.VFC = function RolesScreen() {
...
+        </div>
+        {isEmpty ? (
+          isSearch ? (
+            <MessageBar className={cn(styles.widget, styles.message)}>
+              <FormattedMessage id="RolesScreen.empty.search" />
+            </MessageBar>
```
Comment (tung2744) [inline portal/src/graphql/adminapi/RolesScreen.tsx:176]: Where did you find this design? IMO it should be displayed in the detail list.

File: pkg/admin/facade/rolesgroups.go (hunk 1 above)
Comment (Ngkaokis) [reply pkg/admin/facade/rolesgroups.go:89]: It's mainly for counting the total number. Do you think it's better to add it?

File: portal/src/graphql/adminapi/RolesScreen.tsx (hunk 4 above)
Comment (Ngkaokis) [reply portal/src/graphql/adminapi/RolesScreen.tsx:163]: I copied it from `UsersScreen` :sosad:

Let me fix it together

File: portal/src/graphql/adminapi/RolesScreen.tsx (hunk 6 above)
Comment (Ngkaokis) [reply portal/src/graphql/adminapi/RolesScreen.tsx:176]: Thx. Missed the below design so followed the old design of users list

**Comment by Ngkaokis** [comment]
> It seems the table is too long in narrow screens.

Let me just randomly adjust the min/max width of the columns to make it look fine now may need designer to make a version on mobile

cc @fungc-io 

> There shouldn't be horizontal scrollbar on the outer window. I suggest you to set "Show scroll bars" to "Always" in your system preference, so you will be easier to spot this kind of problem.

Not sure if I set it correct, but I can't see the scroll bar you see

Screenshot 2024-03-11 at 12 50 00 PM

File: portal/src/graphql/adminapi/RolesScreen.tsx (hunk 6 above)
Comment (Ngkaokis) [reply portal/src/graphql/adminapi/RolesScreen.tsx:176]: Not so sure how to handle the empty state in mobile view? :sosad:

Expect it to be static or scroll with columns?

https://github.com/authgear/authgear-server/assets/63200569/f2370bc9-ab27-4ca8-967e-01eed44a51c1

**Comment by tung2744** [comment]
> Not sure if I set it correct, but I can't see the scroll bar you see

@Ngkaokis 

I tried in your latest commit, it seems fixed.

File: portal/src/graphql/adminapi/RolesScreen.tsx (hunk 6 above)
Comment (tung2744) [reply portal/src/graphql/adminapi/RolesScreen.tsx]: If scroll with columns, there is a chance that user cannot see it.
But static looks a bit weird too. Maybe @fungc-io What do you think?

File: pkg/admin/facade/rolesgroups.go (hunk 1 above)
Comment (tung2744) [reply pkg/admin/facade/rolesgroups.go:89]: @Ngkaokis I think pagination will fail if the total count doesn't respect the search keyword?

Comment (Ngkaokis) [reply pkg/admin/facade/rolesgroups.go:89]: If we stick we the behaviour as users list, searching should not show pagination and would query all results (although we limit the search result to 10 in users list for performance issue. But I think we don't have issue with roles and groups query)

Comment (tung2744) [reply pkg/admin/facade/rolesgroups.go:89]: Oh I see. Lets keep it return null if search keyword exist first to be consistent with user query. And lets confirm it with @fungc-io later.

**Comment by tung2744** [comment]
@Ngkaokis Sorry the scrollbar problem still exist, please try a narrow screen size (but not mobile), and add more rows.
Screenshot 2024-03-11 at 2 01 10 PM

File: portal/src/graphql/adminapi/RolesScreen.module.css [hunk 7]
```diff
@@ -1,12 +1,19 @@
 .content {
-  @apply mt-11;
+  flex: 1 1 0;
```
Comment (tung2744) [inline portal/src/graphql/adminapi/RolesScreen.module.css:2]: To fix the scrollbar problem, it seems you need overflow: hidden; here

File: portal/src/graphql/adminapi/RolesScreen.tsx [hunk 8]
```diff
@@ -64,34 +86,83 @@ const RolesScreen: React.VFC = function RolesScreen() {
...
+        </div>
+        <SearchBox {...searchBoxProps} />
       </div>
       {isEmpty ? (
         <RolesScreenEmptyState className={styles.widget} />
       ) : (
-        // TODO: implement list view
-        <RolesScreenEmptyState className={styles.widget} />
+        <RolesList
+          className={styles.widget}
```
Comment (tung2744) [inline portal/src/graphql/adminapi/RolesScreen.tsx:157]: To fix the scrollbar problem, it seems you need min-height: 0 here

Comment (Ngkaokis) [reply portal/src/graphql/adminapi/RolesScreen.tsx:157]: overflow: auto seems more intuitive to me

File: portal/src/graphql/adminapi/RolesScreen.module.css (hunk 7 above)
Comment (Ngkaokis) [reply portal/src/graphql/adminapi/RolesScreen.module.css:2]: I think should just add `overflow: auto;` here

Comment (tung2744) [reply portal/src/graphql/adminapi/RolesScreen.module.css:2]: No, because the scrollbar should appear in content of CommandBarContainer.

**Comment by fungc-io** [comment]
@Ngkaokis @tung2744 
When no result,
Hiding the columns and show a message bar like this in the "User" list seems looks better imo
image

**Comment by tung2744** [comment]
@fungc-io Then other lists should follow this design too?
Such as https://www.figma.com/file/bhUac10bqlsRVXF5HiwCjW/Q1-%2F-2024?type=design&node-id=42-5664&mode=design&t=iFDAUZQOpvTa85MR-4

**Comment by fungc-io** [comment]
@tung2744 
yes, like the audit log,
image

it looks better on all screen sizes
Thanks for pointing out the design problem 🙇

**Pull request merged**